### PR TITLE
pass all message args to output channels

### LINKED ISSUE #3057: Telegram button options do not work.
inline, custom and vertical button options are not passed on from telegram.py to channel.py

### PR DESCRIPTION
**Pull request by erohmensing**
**Proposed changes**:
- pass all message args to output channels for further processing in output channel methods
- fix #3057

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation - **migration needed** - custom output channels need to add `**kwargs` to the parameters of any subclassed send methods
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

File: rasa/core/events/__init__.py [hunk 1]
```diff
@@ -307,9 +307,10 @@ class BotUttered(Event):
 
     type_name = "bot"
 
-    def __init__(self, text=None, data=None, timestamp=None):
+    def __init__(self, text=None, data={}, metadata={}, timestamp=None):
```
Comment (tmbo) [inline rasa/core/events/__init__.py]:
```suggestion
    def __init__(self, text=None, data=None, metadata=None, timestamp=None):
```
default values should *never* be `[]` or `{}`. the dict / array will only be created *once* and then reused for avery future call. this means if you then do `data["key"] = "value"` every future call of init will not have `{}` as a default value but `{"key": "value"}`.

File: rasa/core/events/__init__.py [hunk 2]
```diff
@@ -307,9 +307,10 @@ class BotUttered(Event):
...
         self.text = text
-        self.data = data or {}
+        self.data = data
```
Comment (tmbo) [inline rasa/core/events/__init__.py]:
```suggestion
        self.data = data or {}
```

File: rasa/core/events/__init__.py [hunk 3]
```diff
@@ -307,9 +307,10 @@ class BotUttered(Event):
...
+        self.metadata = metadata
```
Comment (tmbo) [inline rasa/core/events/__init__.py]:
```suggestion
        self.metadata = metadata or {}
```

File: rasa/core/events/__init__.py [hunk 4]
```diff
@@ -347,13 +348,8 @@ def message(self) -> Dict[Text, Any]:
 
         m = self.data.copy()
         m["text"] = self.text
+        m.update(self.metadata.copy())
```
Comment (tmbo) [inline rasa/core/events/__init__.py]: copy isn't necessary, it will create a shallow copy anyways which we don't need

```suggestion
        m.update(self.metadata)
```

File: rasa/core/events/__init__.py [hunk 5]
```diff
@@ -347,13 +348,8 @@ def message(self) -> Dict[Text, Any]:
...
 
-        if m.get("image") == m.get("attachment"):
```
Comment (tmbo) [inline rasa/core/events/__init__.py]: why did you remove this? removing this should break the tests

File: rasa/core/events/__init__.py (hunk 1 above)
Comment (erohmensing) [reply rasa/core/events/__init__.py]: Good to know, I was wondering about that!

File: rasa/core/events/__init__.py (hunk 4 above)
Comment (erohmensing) [reply rasa/core/events/__init__.py]: Got it -- realized we only do it the first time because we're creating m 👍

File: rasa/core/events/__init__.py (hunk 5 above)
Comment (erohmensing) [reply rasa/core/events/__init__.py]: In the current method, if there's no attachment, then image gets put as both image and attachment, and then we delete the image from attachment here. 

I believe I took care of this [here](https://github.com/RasaHQ/rasa/pull/3423/files#diff-d2d909f773bca52743b6b9106d3184a5R135) -- If there's no attachment and an image then the image will get popped into image and therefore not put into attachment.

Comment (tmbo) [reply rasa/core/events/__init__.py]: ok i see, that won't work: for legacy reasons we *want* to set the image as the attachement (having the same value for both if there is no other attachament).  Let's keep this part here and change the code you linked to
```
            "elements": message.pop("elements", None),
            "quick_replies": message.pop("quick_replies", None),
            "buttons": message.pop("buttons", None),
            # for legacy / compatibility reasons we need to set the image to be the attachement if there
            # is no other attachement (the `.get` is intentional - no `pop` as we still need the `image` 
            # property to set it in the following line)
            "attachment": message.pop("attachment", None) or message.get("image", None),
            "image": message.pop("image", None),
            "custom": message.pop("custom", None),
```

Comment (erohmensing) [reply rasa/core/events/__init__.py]: right, but why are we setting the image as the attachment just to get rid of it later? is it just bc there's a processing step in between that uses attachment that i'm missing? it didn't seem to break the tests

Comment (erohmensing) [reply rasa/core/events/__init__.py]: also happy to do it this way and come back to re-evaluate later, just feel like there's a cleaner way to do this. maybe we `setdefault()` to image and then update with attachment? This would make it so that 

> image as the attachement (having the same value for both if there is no other attachament).

Comment (tmbo) [reply rasa/core/events/__init__.py]: it is not about any code in here, but the `BotUttered` event gets persisted and then used e.g. in the chatui frontend on the webpage which also reads these events. Changing the value of the attachement would break that ui.

Comment (tmbo) [reply rasa/core/events/__init__.py]: this is also why we actualy need to set the value and not just pretend it has this value (using `setdefault`)

Comment (erohmensing) [reply rasa/core/events/__init__.py]: Then doesn't moving the image to `image` instead of `attachment` below also break the BotUttered event?

Comment (erohmensing) [reply rasa/core/events/__init__.py]: Then doesn't moving the image to `image` instead of `attachment` below also break the BotUttered event?

Comment (erohmensing) [reply rasa/core/events/__init__.py]: Also can't we handle this in the `create_bot_utterance()` method?

Comment (tmbo) [reply rasa/core/events/__init__.py]: I am not sure what you mean.

- `create_bot_utterance` needs to create a backwards compatible `BotUttered` event (setting image as attachement if no other attachement is present) 
- this `message` function here (in the lines we are currently commenting on) is used to create the message dict that gets handed over to the output channel. to avoid sending the image twice, we remove it from the attachement if it is set there.

Comment (erohmensing) [reply rasa/core/events/__init__.py]: Ahhhhhh okay. now i'm with it, this is what I meant by this 

> is it just bc there's a processing step in between that uses attachment that i'm missing?

so the persisting of the botmessage needs the attachment to also be the image, but then we delete it before sending it to the output channel. All is cleared up! Will clean this up, let me know if you want to rereview before merge but i'll jsut implement your suggestions

Comment (tmbo) [reply rasa/core/events/__init__.py]: just send me a quick note once you are ready and I'll hit the merge button :+1:
